GCC_ARM: newlib thread safety

adamgreen opened the pull request:
Make the **newlib** Standard C Library shipped in the **GNU ARM Embedded *6* Toolchain** thread safe when used with **mbed-os**.

There are four major components of this PR:
* Switch the syscall functions in **platform/mbed_retarget.cpp** to be *reentrant*. This adds a **\_r** extension to the syscall function name and an additional **struct \_reent** parameter. This \_reent structure contains per thread copies of important C Library state such as **errno**.
* Modify RTX to update **\_impure_ptr** on each thread context switch. \_impure_ptr is switched to point to the \_reent structure for the currently running thread.
* Implement **\_\_retarget\_lock** newlib interface in mbed-os. The full version of newlib shipped with the latest GNU ARM Embedded Toolchain calls this generic locking interface when it needs exclusive access to C Library globals such as the heap (ie. malloc()/free()).
* Even though newlib-nano was built single threaded, there are still many areas of the code which call locking routines that can be provided to synchronize key areas. I have added implementations of these locking routines to mbed-os when linking with newlib-nano. I have also used GNU linker wraps for some newlib functions which are missing the required lock calls in newlib-nano.

## Important Things To Document
* The libspace pool only has room for **OS_THREAD_LIBSPACE_NUM** reentrancy structures, similar to the existing ARM implementation. For the ARM Standard C Library, a reentrancy structure is only required for threads that actually use the Standard C Library and end up calling **\_\_user_perthread\_libspace()**. For the newlib Standard C Library, a structure is required for every thread that executes, even if they never use the Standard C Library. This means that the original default of OS_THREAD_LIBSPACE_NUM=4 is too small on GCC_ARM. I increased this number to 5 so that a simple TCP/IP server application could still run after these changes. It might make sense to increase this number even more for GCC_ARM. For newlib-nano, each additional thread added increases the RAM usage by 100 bytes but for the full version of newlib it increases it by more than 1000 bytes. It should be documented that the user will need to increase this setting in **rtos/rtx5/TARGET_CORTEX_M/RTX_Config.h** if they hit a **"CMSIS-RTOS error: STD C/C++ library libspace not available..."** stop message. It needs to match the number of additional worker threads created in their application (explicitly created by the developer and implicitly by libraries such as the networking stack). The count doesn't need to include the main thread itself as a \_reent structure is always reserved for it separately.
* The per **FILE\*** locks are the only area that can't be made thread safe with **newlib-nano** after applying this PR. It will need to be documented that multiple threads sharing the same FILE\* object need to manually synchronize their usage when using newlib-nano. The full version of newlib does contain the required synchronization for these locks.
* These newlib Standard C Library thread safety changes only apply when you are using recent releases of the GCC_ARM tools, **GNU ARM Embedded 6** or newer.

## Other Important Notes/Questions
* This PR doesn't include the necessary changes to the mbed build system to add the wraps which are required when linking with **newlib-nano**. It should be noted that no additional wraps are required when linking with the full version of newlib. I conducted my testing with custom makefiles that included these wraps. The additional wraps required for the mbed build tools when linking with newlib-nano are:
  * ```--wrap=arc4random,```
  * ```--wrap=arc4random_buf,```
  * ```--wrap=__sfp,```
  * ```--wrap=__sinit,```
  * ```--wrap=at_quick_exit```
* I tested these changes with the GNU ARM Embedded 6 Toolchain on macOS using custom makefiles against the following 3 target devices:
  * KL25Z (Cortex-M0+)
  * LPC1768 (Cortex-M3)
  * K64F (Cortex-M4F)
* I also uploaded this change to the online compiler and tested there as well using my LPC1768 board to find and correct any build breaks that I introduced for the ARM compiler.

Please refer to individual commit messages in this PR for more detailed information about each change.

0xc0170 commented:
cc @c1728p9 @kjbracey-arm @bulislaw

bulislaw commented:
Please remove changes to RTX and submit them separately to CMSIS/RTX repo. We are trying not to modify RTX copy in mbed OS unless it's absolutely necessary. https://github.com/ARM-software/CMSIS_5/

tommikas commented:
I'm not sure if you're aware already, but just in case...

jenkins/pr-head is failing at runtime with "CMSIS-RTOS error: STD C/C++ library libspace not available (status: 0x4, task ID: 0x0)" when using Tickers.

0xc0170 commented:
@adamgreen Thanks for this work ! 👍 

> jenkins/pr-head is failing at runtime with "CMSIS-RTOS error: STD C/C++ library libspace not available (status: 0x4, task ID: 0x0)" when using Tickers.

Seems like that number is not enough. How can we fix this. Is this only allowed to be set as static, no dynamic options? 

Commit 594407d could be sent upstream to CMSIS5 , would be nice to share it there as it improves gdb experience.

Commit e3d4176 however is related to our boot sequence, isn't it? Can it be extracted and proposed to CMSIS5 repository ?

c1728p9 commented:
Thanks for the PR, this is great work! 

To follow on @0xc0170 comments, would it be possible to break this PR into several smaller ones? More detailed feedback could then be given on each change. You could probably use each commit message as a PR description since they contain a lot of valuable information.

A few bits of feedback regarding this PR - Could the malloc locks be left in? That way there won't be regressions for people using older versions of the GCC tools. Alternatively maybe this code could be switched out when building with GCC 6 if there is concern with the performance overhead of the extra locks.

Finally, losing the ability to have an arbitrary number of threads at run-time is a fairly big downside to this change. Its also less than ideal that the task switch handler needs to call "__user_perthread_libspace" to lookup the _reent structure. What if rather than looking it up, the _reent pointer was stored either in a TCB or on the stack (similar to a register) during context switch? That way lookups would take the same amount of time regardless of number of threads. Furthermore, this would allow a thread to change it's _reent pointer at run-time, since on the next context switch the new value would be saved. The initial _reent pointer on thread creation could point to a dummy _reent pointer (or maybe NULL). The real _reent pointer could then be dynamically allocated and set just before control is passed to the thread's real entry point. What are your thoughts on this @adamgreen?

adamgreen commented:
@bulislaw 
> Please remove changes to RTX and submit them separately to CMSIS/RTX repo. We are trying not to modify RTX copy in mbed OS unless it's absolutely necessary. https://github.com/ARM-software/CMSIS_5/

I have removed the 594407d5c123f5716dd5e18f2184d0532075e477 commit from the PR since it was a GDB specific change to RTX. The other changes to RTX are part of commit e3d4176345871f8ba16ec5e9b5182688690ef78c and they are mbed-os specific, with some code changes in the RTX context switching code and the rest in rtos/mbed_boot.c

@tommikas
> I'm not sure if you're aware already, but just in case...

> jenkins/pr-head is failing at runtime with "CMSIS-RTOS error: STD C/C++ library libspaxce not available (status: 0x4, task ID: 0x0)" when using Tickers.

I was aware that it had failed but I don't seem able to see the actual error text since when I click on the details link, I just get an error indicating that I can't connect to **jenkins-internal.mbed.com**. Is this the only test that failed? If so then that is a lot better than I expected. It's failure is related to the **Important Things To Document** comment I made in the original PR message about **OS_THREAD_LIBSPACE_NUM**. That macro needs to be increased for that particular test. I wouldn't increase it anymore in general since for full newlib, it adds >1000 bytes for each additional thread entry.

@0xc0170 
> Seems like that number is not enough. How can we fix this. Is this only allowed to be set as static, no dynamic options?

Yes, it should be increased for this test and any other test which hit similar **"CMSIS-RTOS error: STD C/C++ library libspace not available..."** stop messages. I don't see an easy way to dynamically allocate them since the code only learns that there aren't enough from within an ISR (actually during thread context switching) so that rules out calls to malloc.

> Commit 594407d could be sent upstream to CMSIS5 , would be nice to share it there as it improves gdb experience.

Agreed. I have removed this commit from the PR.

> Commit e3d4176 however is related to our boot sequence, isn't it? Can it be extracted and proposed to CMSIS5 repository ?

Do you have recommendations on how to do this? I started making this code specific to RTX/newlib only with no knowledge of mbed-os specifics but I hit an issue when I wanted to know if the current thread was the main thread and I didn't see an easy to do that from within RTX. I want to short-circuit and use `_global_impure_ptr` for the main thread's reentrancy data since it is already allocated by newlib. Otherwise I would be wasting this 96 bytes for newlib-nano and 1064 bytes for newlib-full.

@c1728p9 
> To follow on @0xc0170 comments, would it be possible to break this PR into several smaller ones? More detailed feedback could then be given on each change. You could probably use each commit message as a PR description since they contain a lot of valuable information.

Anyone who wants to review it a commit at a time can click the **Commits** tab at the top of the PR to drill down and look at the commit specific commits and changes. I have conducted the bulk of my testing with the commits together as they are in the PR. Splitting up the commits just increases my test hit and increases the chances of errors creeping in if some commits are taken and others aren't. I can remove commits like I have already done but only ones I don't plan to resubmit to mbed-os separately at a later point in time (will instead push to a different repo or just drop all together.)

> A few bits of feedback regarding this PR - Could the malloc locks be left in? That way there won't be regressions for people using older versions of the GCC tools. Alternatively maybe this code could be switched out when building with GCC 6 if there is concern with the performance overhead of the extra locks.

I would prefer to use the malloc_un/lock() routines as they are implemented in newlib-full and not pull them unnecessarily into the mbed-os code base. If a user wants newlib thread safety then they should switch to GCC 6+ where it is actually supported. The locks are switched out if you use the single threaded mbed 2.0 SDK.

> Finally, losing the ability to have an arbitrary number of threads at run-time is a fairly big downside to this change. Its also less than ideal that the task switch handler needs to call "\_\_user_perthread_libspace" to lookup the \_reent structure. What if rather than looking it up, the \_reent pointer was stored either in a TCB or on the stack (similar to a register) during context switch? That way lookups would take the same amount of time regardless of number of threads. Furthermore, this would allow a thread to change it's \_reent pointer at run-time, since on the next context switch the new value would be saved. The initial \_reent pointer on thread creation could point to a dummy \_reent pointer (or maybe NULL). The real \_reent pointer could then be dynamically allocated and set just before control is passed to the thread's real entry point.

I used this approach because it is based on [how RTX implements it for the ARM compiler](https://github.com/ARMmbed/mbed-os/blob/master/rtos/rtx5/TARGET_CORTEX_M/rtx_lib.c#L543) and I wanted to minimize the changes to RTX. The best thing might be to just pull the per thread memory for the Standard C Library back into the [thread control block](https://github.com/ARMmbed/mbed-os/blob/master/rtos/rtx5/TARGET_CORTEX_M/rtx_os.h#L105) like was [done with an older version of RTX](https://github.com/adamgreen/gcc4mbed/blob/gcc5.0-2015-q4_mbed119/external/mbed/libraries/rtos/rtx/TARGET_CORTEX_M/os_tcb.h#L41) so that every thread created also has its data allocated for the C lib as well.

Thanks for the feedback!

bulislaw commented:
> I have removed the 594407d commit from the PR since it was a GDB specific change to RTX. The other changes to RTX are part of commit e3d4176 and they are mbed-os specific, with some code changes in the RTX context switching code and the rest in rtos/mbed_boot.c

I don't think we should have any mbed specific changes to RTX/CMSIS, we have few right now to bridge the differences in build system mostly, plus some more for uVisor, which is more complicated. I wouldn't like to add any more. It makes my life harder with updates and it sets wrong precedence. I would say all changes should be discussed and agreed with the CMSIS/RTX team. If they decide they don't belong to RTX/CMSIS we shouldn't be adding them on top of it in mbed.

> I was aware that it had failed but I don't seem able to see the actual error text since when I click on the details link, I just get an error indicating that I can't connect to jenkins-internal.mbed.com. Is this the only test that failed? If so then that is a lot better than I expected. It's failure is related to the Important Things To Document comment I made in the original PR message about OS_THREAD_LIBSPACE_NUM. That macro needs to be increased for that particular test. I wouldn't increase it anymore in general since for full newlib, it adds >1000 bytes for each additional thread entry.

That sounds like another rabbit hole, similar to what we used to have with ARM toolchain and the 'dynamic' mutexes it uses for thread safety. Initially we've allocated some arbitrary number of locks based on our testing. For next months we had to deal with bugs for use cases that happened to require more.  In the end we solved it properly to handle all cases. Is there a way of addressing that so it fits/scales to all possible cases?

> Do you have recommendations on how to do this? I started making this code specific to RTX/newlib only with no knowledge of mbed-os specifics but I hit an issue when I wanted to know if the current thread was the main thread and I didn't see an easy to do that from within RTX. I want to short-circuit and use _global_impure_ptr for the main thread's reentrancy data since it is already allocated by newlib. Otherwise I would be wasting this 96 bytes for newlib-nano and 1064 bytes for newlib-full.

How about we expose return value of creating the main thread as a global variable? So you can compare it later with current thread id?
https://github.com/ARMmbed/mbed-os/blob/e3d4176345871f8ba16ec5e9b5182688690ef78c/rtos/mbed_boot.c#L319

> I used this approach because it is based on how RTX implements it for the ARM compiler and I wanted to minimize the changes to RTX. The best thing might be to just pull the per thread memory for the Standard C Library back into the thread control block like was done with an older version of RTX so that every thread created also has its data allocated for the C lib as well.

It's not true anymore for ARM compiler, it was something that was annoying me for a long time and it's fixed now. We allocate maximum number of static locks (a static memory) and then compliment that with malloced mutexes for any additional dynamic locks.

adamgreen commented:
Thanks for all of the feedback.

It sounds like the best course of action is for you guys to work with the CMSIS/RTX team to come up with a solution that you guys like. I don't know that there is much more I can contribute at this point.

adamgreen commented:
> Finally, losing the ability to have an arbitrary number of threads at run-time is a fairly big downside to this change.

This is also true for the online compiler as well. If you spin up 4 threads, and do something simple like set "errno = 0;" from all of these threads and the main thread, you will hit the same stop message. The problem is just more easily hit on GCC/newlib since all threads, even if they don't call the C runtime need to have library space ready for them since it isn't just requested on demand.
